### PR TITLE
`Mapping` creation and indexing

### PR DESCRIPTION
**Pull request by arnaudgolfouse**
I'm not a fan of the naming of `unknown`, but I don't really have a better idea.

Also this implies that logic code may create values ex nihilo. I assumed this was what we wanted ; but should we rather have a function
```rust
#[logic]
#[trusted]
fn conjure<T>() -> T { dead }
```
?

### CODE REVIEW

**Comment by jhjourdan** [comment]
Indexing: LGTM.

Creation of an arbitrary mapping: why cannot we use `cst` in your specific case?

In general, I think a function like your `conjure` is a good idea. I think there is no reason this should be restricted to `Mapping`. Actually, I think we should even go further, by having the equivalent of an epsilon function in Creusot, which would essentially let us eliminate existential quantifiers in ghost code.

**Comment by arnaudgolfouse** [comment]
> Creation of an arbitrary mapping: why cannot we use cst in your specific case?

Because the type of values of that mapping is a generic type `T`, so I cannot create a value of type `T` (unless of course we add `conjure`).

> In general, I think a function like your `conjure` is a good idea.

I'll add it then! We can think about an epsilon function later, but yes it is something that I wanted occasionally.

**Comment by arnaudgolfouse** [comment]
Btw what is our position on uninhabited types?

**Comment by jhjourdan** [comment]
> Btw what is our position on uninhabited types?

Smt solvers and why3 don't support uninhabited types, they assume that all types are inhabited. It is rather heavy to encode uninhabited types in these tools, so I prefer not supporting them either.

This is the main reason why we don't encode type invariants as subset types.

**Comment by jhjourdan** [comment]
> I'll add it then! We can think about an epsilon function later, but yes it is something that I wanted occasionally.


Why waiting for epsilon ? I don't expect it to be much more difficult than `conjure`, and I'd prefer not having two redundant features in the future.

File: creusot-contracts/src/logic/mapping.rs [hunk 1]
```diff
@@ -24,30 +24,53 @@ impl<A: ?Sized, B: ?Sized> Mapping<A, B> {
...
+    /// This is also called the "epsilon operator": its goal is not extract from `∃x. P(x)`
+    /// a `x` satisfying `P`.
+    ///
+    /// This is the method version of [`such_that`](crate::util::such_that).
+    #[logic]
+    #[open]
+    pub fn such_that(self) -> T {
+        crate::util::such_that(self)
+    }
+}
```
Comment (jhjourdan) [inline creusot-contracts/src/logic/mapping.rs:66]: Quite frankly, I am not convinced there is a need for this. This is redundant, and it does not allow to write shorter code.

**Pull request merged**